lib/new-admin/graphql/modules/authentication.js

```diff
@@ -0,0 +1,300 @@
+    })
+    .then(([_, user, secret, otpauth]) => {
+      return { user_id: user.id, secret, otpauth }
+    })
+    .catch(err => console.error(err))
+}
+
+const deleteSession = (sessionID, context) => {
+  destroySessionIfBeingUsed(sessionID, context)
+  return sessionManager.deleteSessionById(sessionID)
+}
+
+const login = (username, password) => {
+  return authenticateUser(username, password).then(user => {
+    const twoFASecret = user.twofa_code
+    return twoFASecret ? 'INPUT2FA' : 'SETUP2FA'
+  })
+}
+
+const input2FA = (username, password, rememberMe, code, context) => {
+  return authenticateUser(username, password)
+    .then(user => {
+      const secret = user.twofa_code
+      const isCodeValid = otplib.authenticator.verify({ token: code, secret: secret })
+      if (!isCodeValid) throw new authErrors.InvalidTwoFactorError()
+
+      initializeSession(context, user, rememberMe)
+      return true
+    })
+}
+
+const setup2FA = (username, password, rememberMe, secret, codeConfirmation, context) => {
+  if (!secret) throw new authErrors.InvalidTwoFactorError()
+
+  const isCodeValid = otplib.authenticator.verify({ token: codeConfirmation, secret: secret })
+  if (!isCodeValid) throw new authErrors.InvalidTwoFactorError()
+
+  return authenticateUser(username, password)
+    .then(user => {
+      if (user.temp_twofa_code !== secret) {
+        throw new authErrors.InvalidTwoFactorError()
+      }
+
+      initializeSession(context, user, rememberMe)
+      return users.save2FASecret(user.id, secret)
+    })
+    .then(() => true)
+}
+
+const changeUserRole = (code, id, newRole, context) => {
+  const action = (id, newRole) => users.changeUserRole(id, newRole)
+
+  return users.getUserById(id)
+    .then(user => {
+      if (user.role !== 'superuser') {
+        return action(id, newRole)
+      }
+    
+      return confirm2FA(code, context)
+    })
+    .then(() => action(id, newRole))
+}
+
+const enableUser = (code, id, context) => {
+  const action = id => users.enableUser(id)
+
+  return users.getUserById(id)
+    .then(user => {
+      if (user.role !== 'superuser') {
+        return action(id)
+      }
+    
+      return confirm2FA(code, context)
+    })
+    .then(() => action(id))
+}
+
+const disableUser = (code, id, context) => {
+  const action = id => users.disableUser(id)
+
+  return users.getUserById(id)
+    .then(user => {
+      if (user.role !== 'superuser') {
+        return action(id)
+      }
+    
+      return confirm2FA(code, context)
+    })
+    .then(() => action(id))
+}
+
+const createResetPasswordToken = (code, userID, context) => {
+  const action = user => users.createAuthToken(user.id, 'reset_password')
+
+  return users.getUserById(userID)
+    .then(user => {
+      if (user.role !== 'superuser') {
+        return action(user.id)
+      }
+    
+      return confirm2FA(code, context)
+        .then(() => action(user))
+    })
+}
```

Comment:
Now this looks like something that can get extracted. We basically have the exact same logic for a lot of functions here.